src/socket.rs

```diff
@@ -0,0 +1,383 @@
+            join_handle: Mutex::new(Some(join_handle)),
+        }))
+    }
+
+    pub fn url(&self) -> Arc<Url> {
+        self.url.clone()
+    }
+
+    /// Connects this client to the configured Phoenix Channels endpoint
+    ///
+    /// This function must be called before using the client to join channels, etc.
+    ///
+    /// A join handle to the socket worker is returned, we can use this to wait until the worker
+    /// exits to ensure graceful termination. Otherwise, when the handle is dropped, it detaches the
+    /// worker from the task runtime (though it will continue to run in the background)
+    pub async fn connect(&self, timeout: Duration) -> Result<(), ConnectError> {
+        let (connected_sender, connected_receiver) = oneshot::channel();
+        let created_at = Instant::now();
+        let deadline = Instant::now() + timeout;
+
+        self.state_command_sender
+            .send(StateCommand::Connect(Connect {
+                created_at,
+                timeout,
+                connected_sender,
+            }))
+            .await?;
+
+        time::timeout_at(deadline, connected_receiver).await??
+    }
+
+    /// Disconnect the client, regardless of any outstanding channel references
+    ///
+    /// Connected channels will return `ChannelError::Closed` when next used.
+    ///
+    /// New channels will need to be obtained from this client after `connect` is
+    /// called again.
+    pub async fn disconnect(&self) -> Result<(), DisconnectError> {
+        let (disconnected_sender, disconnected_receiver) = oneshot::channel();
+
+        self.state_command_sender
+            .send(StateCommand::Disconnect {
+                disconnected_sender,
+            })
+            .await?;
+
+        disconnected_receiver.await.map_err(From::from)
+    }
+
+    /// Propagates panic from [Listener::listen]
+    pub async fn shutdown(self) -> Result<(), ShutdownError> {
+        self.state_command_sender
+            .send(StateCommand::Shutdown)
+            .await
+            // if already shut down that's OK
+            .ok();
+
+        self.listener_shutdown().await
+    }
+
+    /// Creates a new, unjoined Phoenix Channel
+    pub async fn channel(
+        self: &Arc<Self>,
+        topic: &str,
+        payload: Option<Payload>,
+    ) -> Result<Arc<Channel>, ChannelError> {
+        let (sender, receiver) = oneshot::channel();
+
+        match self
+            .channel_spawn_sender
+            .send(ChannelSpawn {
+                socket: self.clone(),
+                topic: topic.to_string(),
+                payload,
+                sender,
+            })
+            .await
+        {
+            Ok(()) => receiver.await.map(Arc::new).map_err(From::from),
+            Err(_) => Err(self.listener_shutdown().await.unwrap_err().into()),
+        }
+    }
+
+    pub(crate) async fn join(
+        &self,
+        topic: Topic,
+        join_reference: JoinReference,
+        payload: Arc<Payload>,
```

Comment:
This function is only called the `channel::Listener`.  There's no need for it to be generic as the `Socket::channel` has already handled any generics when storing the payload for the channel.

Reply:
Changing the signature of `Socket::channel` to 

```rust
    pub async fn channel<T, P>(
        self: &Arc<Self>,
        topic: T,
        payload: Option<P>,
    ) -> Result<Arc<Channel>, ChannelError>
    where
        T: Into<SharedStr>,
        P: Into<Payload>
```   

Caused the tests to need type ascription for the async tests, so I don't think it's worth it if callers need to give the type for `None`.

```
 let receiver_channel = receiver_client.channel(&topic, None).await.unwrap();
    |                                                            ^^^^ cannot infer type for type parameter `T` declared on the enum `Option`
```

Reply:
> It might make sense for Payload to internally be Payload::Value(Arc) and Payload::Binary(Bytes). Bytes is better than Arc<Vec> for when cheap cloning is desired.
>
> https://docs.rs/bytes/1.4.0/bytes/struct.Bytes.html

✅